pkg/cincinnati/cincinnati.go

```diff
@@ -0,0 +1,120 @@
+package cincinnati
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/blang/semver"
+	"github.com/google/uuid"
+)
+
+const (
+	ContentTypeGraphV1 = "application/vnd.redhat.cincinnati.graph+json; version=1.0"
+)
+
+type Client struct {
+	id uuid.UUID
+}
+
+func NewClient(uuid uuid.UUID) Client {
+	return Client{
+		id: uuid,
+	}
+}
+
+func (c *Client) GetUpdate(upstream string, version semver.Version, channel string) ([]string, error) {
+	req, err := http.NewRequest("GET", upstream, nil)
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Add("Accept", ContentTypeGraphV1)
+
+	client := http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("bad HTTP status: %s", resp.Status)
+	}
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	var graph graph
+	if err = json.Unmarshal(body, &graph); err != nil {
+		return nil, err
+	}
+
+	var rootIdx int
+	for i, node := range graph.Nodes {
+		if version.EQ(node.Version) {
+			rootIdx = i
+			break
+		}
+	}
+
+	var nextIdxs []int
+	for _, edge := range graph.Edges {
+		if edge.Origin == rootIdx {
+			nextIdxs = append(nextIdxs, edge.Destination)
+		}
+	}
+
+	var payloads []string
+	for _, i := range nextIdxs {
+		payloads = append(payloads, graph.Nodes[i].Payload)
+	}
+
+	return payloads, nil
+}
+
+type graph struct {
+	Nodes []node
+	Edges []edge
+}
+
+type node struct {
+	Version semver.Version
+	Payload string
+}
+
+type edge struct {
+	Origin      int
+	Destination int
+}
+
+func (e *edge) UnmarshalJSON(data []byte) error {
```

Comment:
Why special unmarshal?

Reply:
Because we are going from an array of indices to a struct.

For example, given the following response:

```
{
  nodes: {...},
  edges: [[0, 1]]
}
```

This needs to be unmarshalled into a struct:

```go
{
  Origin: 0,
  Destination: 1,
}
```